src/utils/uri.js

```diff
@@ -166,3 +167,72 @@ var SmsURI = class URI {
     }
 };
 
+
+/**
+ * Thumbnailer class from Polari
+ *
+ * Credits: https://gitlab.gnome.org/GNOME/polari/-/merge_requests/134
+ */
+var Thumbnailer = class Thumbnailer {
+    static getDefault() {
+        if (!this._singleton)
+            this._singleton = new Thumbnailer();
+        return this._singleton;
+    }
+
+    constructor() {
+        this._urlQueue = [];
+        this._subProc = null;
+        this._thumbnailsDir = `${GLib.get_user_cache_dir()}/gsconnect/thumbnails/`;
+
+        GLib.mkdir_with_parents(this._thumbnailsDir, 0o755);
+    }
+
+    getThumbnail(uri, callback) {
+        let filename = this._generateFilename(uri);
+        let data = {uri, filename, callback};
+
+        this._processData(data);
+    }
+
+    _processData(data) {
+        if (GLib.file_test(`${data.filename}`, GLib.FileTest.EXISTS))
+            this._generationDone(data);
+        else if (!this._subProc)
+            this._generateThumbnail(data);
+        else
+            this._urlQueue.push(data);
+    }
+
+    _generationDone(data) {
+        data.callback(data.filename);
+
+        let nextData = this._urlQueue.shift();
+        if (nextData)
+            this._processData(nextData);
+    }
+
+    _generateThumbnail(data) {
+        let {filename, uri} = data;
+        this._subProc = Gio.Subprocess.new(
+            ['gjs', `${gsconnect.extdatadir}/utils/thumbnailer.js`, uri, filename],
```

Comment:
Yeah, we prefer shebangs, as they are easier to patch automatically to a concrete path.

Reply:
Though, why not just import that as a module? It should be possible to make it sync as well, should not it?

Reply:
That's true, I've been playing with inlining it a bit. The catch is gjs still doesn't deal well with garbage collection complex objects like GtkWindow, so so far it has some pretty big memory increases if I run it in-process.

I'll work more at it and ping you again closer to merge-time.